test/adt_test.exs

```diff
@@ -54,6 +54,29 @@ defmodule AdtTest do
     assert AdtDefinitionThree.variants == [AdtTest.AdtDefinitionThree.Baz, AdtTest.AdtDefinitionThree.Bar, AdtTest.AdtDefinitionThree.Foo]
   end
 
+  test "case" do
+    require AdtTest.AdtDefinition
+
+    foo = %AdtDefinition.Foo{}
+
+    result = AdtDefinition.case foo, [
+      Foo: fn(_) -> "1" end,
+      Bar: fn(_) -> "2" end
+    ]
+    assert result == "1"
+  end
+
+  test "non-exhaustive case" do
+    error = catch_error Code.eval_string """
```

Comment:
It's weird but

```elixir
    error = catch_error(Code.eval_quoted do
      require AdtTest.AdtDefinition

      result = AdtTest.AdtDefinition.case %AdtTest.AdtDefinition.Foo{}, [
        Foo: fn(x) -> "foo" end
      ]
    end)
```

Seems to throw an error in the test, and not catch. Something subtle to do with the order the arguments are evaluated maybe?